src/fs.rs

```diff
@@ -0,0 +1,400 @@
+            }
+        })())
+    }
+    fn host_close(&self, fd: u64) -> IOResult<()> {
+        Ok((|| {
+            if unsafe { libc::close(fd as libc::c_int) } == 0 {
+                Ok(())
+            } else {
+                Err(errno())
+            }
+        })())
+    }
+    fn host_pread(&self, fd: u64, count: u64, offset: u64) -> IOResult<Vec<u8>> {
+        Ok((|| {
+            // Allocate `count` (converted to usize) bytes
+            let count = usize::try_from(count).unwrap_or(std::usize::MAX);
+            let mut buf: Vec<u8> = Vec::with_capacity(count);
+
+            // Fill bytes as many as `read` bytes
+            let read: isize = unsafe {
+                libc::pread(
+                    fd as libc::c_int,
+                    buf.as_mut_ptr() as *mut libc::c_void,
+                    count,
+                    offset as libc::off_t,
+                )
+            };
+
+            if read >= 0 {
+                // Mark bytes up to `read` as initiated
+                unsafe {
+                    buf.set_len(read as usize);
+                }
+                Ok(buf)
+            } else {
+                Err(errno())
+            }
+        })())
+    }
+    fn host_pwrite(&self, fd: u64, offset: u64, data: Vec<u8>) -> IOResult<u64> {
+        Ok((|| {
+            // Write data, as much as `written` bytes of data
+            let written: isize = unsafe {
+                libc::pwrite(
+                    fd as libc::c_int,
+                    data.as_ptr() as *const libc::c_void,
+                    data.len(),
+                    offset as libc::off_t,
+                )
+            };
+
+            if written >= 0 {
+                Ok(written as u64)
+            } else {
+                Err(errno())
+            }
+        })())
+    }
+    fn host_fstat(&self, fd: u64) -> IOResult<HostStat> {
+        Ok((|| {
+            let mut stat: MaybeUninit<libc::stat> = MaybeUninit::uninit();
+
+            if unsafe { libc::fstat(fd as libc::c_int, stat.as_mut_ptr()) } == 0 {
+                let stat = unsafe { stat.assume_init() };
+                Ok(HostStat {
+                    st_dev: stat.st_dev as _,
+                    st_ino: stat.st_ino as _,
+                    st_mode: stat.st_mode as _,
+                    st_nlink: stat.st_nlink as _,
+                    st_uid: stat.st_uid as _,
+                    st_gid: stat.st_gid as _,
+                    st_rdev: stat.st_rdev as _,
+                    st_size: stat.st_size as _,
+                    st_blksize: stat.st_blksize as _,
+                    st_blocks: stat.st_blocks as _,
+                    st_atime: stat.st_atime as _,
+                    st_mtime: stat.st_mtime as _,
+                    st_ctime: stat.st_ctime as _,
+                })
+            } else {
+                Err(errno())
+            }
+        })())
+    }
+    fn host_unlink(&self, _filename: Vec<u8>) -> IOResult<()> {
+        Err(())
```

Comment:
Not sure if unlink should be implemented as it could be dangerous, so I left it and a few functions as TODO